### PR TITLE
Add functional `tfexec` client

### PR DESCRIPTION
**Pull request by git-ival**
## Problem
There was no method via the golang test framework to interact with terraform.
 
## Solution
Added a tfexec client which wraps the [`tfexec`](https://github.com/hashicorp/terraform-exec) library to provide easy-to-use wrapper functions around common operations. 
 
## Testing

Did various tests on local machine, not tested in Jenkins.

Adds additional dependencies to go.mod:
- github.com/hashicorp/go-version
- github.com/hashicorp/hc-install
- github.com/hashicorp/terraform-exec
- github.com/hashicorp/terraform-json

Example tfexec config:
```yaml
tfexec:
  workspaceName: "workspaceToApplyModuleIn"
  workingDir: /path/to/a/terraform/module/
  execPath: /path/to/executable/terraform <--- if omitted, there is an InstallTfVersion() function to enable installing a specified terraform version
  varFilePath: /path/to/terraform/module.tfvars
  planFilePath: "/pre/existing/plan/file/to/use"
  planOpts:
    outDir: /path/to/output/planfile/to
```

### CODE REVIEW

**Review by caliskanugur — COMMENTED** [review]
Looks good, I've got a couple of questions and nits

File: clients/tfexec/tfexec.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+func (c *Client) SetupWorkspace(opts ...tfexec.WorkspaceNewCmdOption) error {
+	ws := ""
+	if c.TerraformConfig.WorkspaceName != "" {
+		ws = c.TerraformConfig.WorkspaceName
+	}
+
+	var err error
+	if ws != "" {
+		wsExists, _ := c.WorkspaceExists(ws)
```
[the remark this answers is not in the document]
Comment (git-ival) [reply clients/tfexec/tfexec.go:112]: I'm not sure I understand the question. There are various error states that could occur, I am trying to validate that none of those errors will occur and if any of them are triggered then we return that error on line 124.

File: clients/tfexec/tfexec.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	err := c.Terraform.Init(context.Background(), opts...)
+	if err != nil {
+		return errors.Wrap(err, "InitTerraform: ")
+	}
+	return nil
+}
+
+// InstallTfVersion uses hc-install in order to install the desired Terraform version
+// Returns the execPath of the newly installed Terraform binary
+func InstallTfVersion(tfVersion string) (string, error) {
```
[the remark this answers is not in the document]
Comment (git-ival) [reply clients/tfexec/tfexec.go:58]: The linter doesnt seem to care about the naming here, I can change it if needed

[the remark this answers is not in the document]
Comment (caliskanugur) [reply clients/tfexec/tfexec.go:58]: Looks good to me, thanks so much for checking it!

File: clients/tfexec/tfexec.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (caliskanugur) [reply clients/tfexec/tfexec.go:126]: That part makes sense to me. What I am asking is this: https://github.com/rancher/shepherd/pull/27/files/ec337d8ba8bb125329f0dacef796d7e4e163429e#diff-8bf95bb523d6e89ae62a6c5b3832ae08d5473f3cec688b1a34830b910176004cR91. This function returns error but the return value isn't used here by `_`

**Comment by git-ival** [comment]
@caliskanugur Anything else I should modify before I squash?

File: go.mod [hunk 3]
```diff
@@ -35,7 +39,7 @@ require (
 	github.com/sirupsen/logrus v1.9.3
 	github.com/spf13/cobra v1.7.0
 	github.com/stretchr/testify v1.8.4
-	golang.org/x/crypto v0.14.0
+	golang.org/x/crypto v0.17.0
```
Comment (igomez06) [inline go.mod:42]: Hey Iramis, why is this version being bumped?

File: go.mod [hunk 4]
```diff
@@ -54,14 +58,17 @@ require (
 	github.com/MakeNowJust/heredoc v1.0.0 // indirect
 	github.com/Masterminds/goutils v1.1.1 // indirect
 	github.com/Masterminds/sprig/v3 v3.2.3 // indirect
-	github.com/Microsoft/go-winio v0.5.2 // indirect
+	github.com/Microsoft/go-winio v0.6.1 // indirect
```
Comment (igomez06) [inline go.mod:61]: Same here

**Review by igomez06 — COMMENTED** [review]
questions of version bumps.

File: go.mod (hunk 3 above)
Comment (git-ival) [reply go.mod:42]: Seems like the new hashicorp deps rely on this version

File: go.mod (hunk 4 above)
Comment (git-ival) [reply go.mod:61]: Seems like the new hashicorp deps rely on this version

**Review by igomez06 — APPROVED** [review]
version bumps all look good. Thanks Iramis.

**Comment by igomez06** [comment]
@git-ival can squash your commits, and push up the back ports?

**Review by caliskanugur — COMMENTED** [review]
Looks good to me Iramis, could we get another rebase and a squash?

File: go.mod (hunk 3 above)
Comment (git-ival) [reply go.mod:42]: Discussed separately, resolved

**Review by caliskanugur — APPROVED** [review]
Looks good to me

**Pull request merged**